cluster/cluster.go

```diff
@@ -217,6 +217,9 @@ func (p *Peer) Join(
 	n, err := p.mlist.Join(p.resolvedPeers)
 	if err != nil {
 		level.Warn(p.logger).Log("msg", "failed to join cluster", "err", err)
+		if reconnectInterval != 0 {
```

Comment:
How about returning the error only when reconnectInterval is zero? This way, `main()` can still treat the error as a hard fail.

```go
if err != nil {
    if reconnectInterval == 0 {
        return err
    }
    level.Info(p.logger).Log("msg", fmt.Sprintf("will retry joining cluster in %v", reconnectInterval.String()))
}
```

Reply:
I was thinking about that as well. Good point.

Still I think a function called `Join` should fail if the join action failed.

>  This way, main() can still treat the error as a hard fail

I would prefer Alertmanager not to `exit(1)` on any cluster related failures. A partitioned Alertmanager cluster is still better than no Alertmanager cluster. What do you think @simonpasquier?

Reply:
> A partitioned Alertmanager cluster is still better than no Alertmanager cluster.

I can hear that. It's definitely a balance between being functional and correct. Provided that there's an alert firing when the cluster's size is less than expected, it's indeed better to start in case of a partition.

Reply:
Do you still want me to do any follow up changes in this regards?

Reply:
No I'm fine as it is now.